### PR TITLE
[WS1] Close embedding and LM-head invariance coverage

### LINKED ISSUE #151: [WS1][kernels] Batch-invariant embedding + LM head projection
Part of WS1 — Full Batch-Invariant Forward Chain (epic: #<WS1 tracking issue>)

## Why

The input embedding lookup and the final vocab projection bracket the network. The LM head is a large matmul sitting directly upstream of logprob, so any batch-dependent reduction there lands in the logprobs immediately. The embedding lookup is simpler but must be confirmed not to branch on batch shape. Both must be on the batch-invariant path or the chain is not actually closed.

## Scope

Confirm the input embedding and the LM-head projection run batch-invariantly.

- Confirm the embedding lookup (gather) produces identical vectors for a token regardless of batch size, position, or padding (no shape-dependent path).
- Route the LM-head vocab projection through the batch-invariant GEMM (the matmul issue), with a fixed K-accumulation order over the hidden dimension.
- Confirm tied-embedding weight sharing (if used by the model) does not change the reduction path between the two.
- Validate both against the #108 harness across the standard sweep, with particular attention to the LM-head -> logprob handoff.

## Out of scope

- The GEMM kernel itself (covered by the matmul issue; this issue consumes it).
- The logprob reduction (covered by the logprob issue; this issue feeds it).
- Vocab-parallel distributed LM head; weight-tying policy changes.
- FP8.

## Acceptance criteria

- Embedding output for a fixed token is bitwise-identical across all batch configs in the sweep.
- LM-head logits for a fixed position are identical (within #108 tolerance) across batch=1/N, chunked-prefill on/off, padding layouts.
- The LM-head -> logprob path is verified jointly: the logits handed to logprob do not drift with batch shape.
- Embedding / LM-head backward paths (incl. tied-weight gradient accumulation, if used) pass the shared gradient-invariance check from the WS1 backward-consistency issue.
- Both ops pass the #108 shared test helper.

## Notes

- Depends on #108 and on the batch-invariant matmul issue (the LM head routes through it).
- Pairs tightly with the logprob issue — drift here is the most direct cause of logprob drift, so validate the two together.

## Planned PRs

- [ ] Embedding-lookup invariance tests (fixed token/position -> identical vector)
- [ ] Route the LM-head vocab projection through the deterministic GEMM
- [ ] Tied-weight sharing consistency check (if the model uses tied embeddings)
- [ ] LM-head -> logprob handoff test
- [ ] Embedding / LM-head backward invariance via the shared gradient check; wire through #108 harness

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary
- close the remaining issue #151 coverage gap by adding RL-batch embedding invariance tests for fixed probe tokens, row permutation, padding mutation, and embedding backward layout invariance
- make the native LM-head reference use row-wise fixed-K projection so logits are bitwise batch-invariant, and update the LM-head docs/tests accordingly
- wire embedding and lm_head into the #108 operator spec path, including shared helper gradient checks
- make DeepSpeed linear_logp backend lookup pass the explicit worker device instead of relying on global device_ctx

## Relationship to existing PRs
- Supersedes the useful coverage intent from #190 without carrying its merge conflicts.
- Completes the #194 device-routing/review gap on top of current main without carrying its stale failing diff.

Closes #151

## Validation
- PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest -q tests/test_issue151_embedding_lm_head_invariance.py tests/test_embedding.py tests/test_lm_head.py tests/test_linear_logp.py tests/test_op_checks.py tests/test_operator_inputs.py tests/test_deepspeed_training_worker.py rl_engine/tests/test_dispatch.py
- python -m black --target-version py310 --line-length=100 --check <changed python files>
- python -m isort --profile black --line-length=100 --check-only <changed python files>
- python -m ruff check <changed python files>
- python -m mypy --python-version 3.10 --ignore-missing-imports rl_engine/
- python -m pre_commit run --files <changed files>

## Summary by CodeRabbit

* **New Features**
  * Added SM90 (Hopper) optimized backends for embedding and LM-head, including explicit `forward_fp32` variants.
  * Extended kernel/operator coverage and dispatch to handle `embedding` and `lm_head` with device-aware backend selection.
* **Bug Fixes**
  * Improved LM-head projection to enforce stable fixed-K row-wise behavior, strict fp32 semantics, and correct zero-row output handling.
  * Ensured embedding and LM-head outputs/gradients remain invariant across batch layouts, row permutations, and padded-token tail mutations.
* **Documentation**
  * Updated embedding and LM-head operator contracts, dtype/precision behavior, and backend dispatch guidance.
* **Tests**
  * Added/expanded invariance, dispatch/device-selection, SM90 wrapper, and fixed-K correctness regression coverage.

### CODE REVIEW

**Comment by maxiaosong1124** [comment]
please resolve conflicts first, thanks!

**Comment by maxiaosong1124** [comment]
@inaniloquentee I only see the PyTorch implementation; I don’t see the corresponding CUDA and Triton implementations. If they haven't been implemented yet, could you please add them? Thank you.

**Comment by Flink-ddd** [comment]
cc @maxiaosong1124 @KJLdefeated @frank-2077 PTAL, we need merge this pr tonight. Thank you.

File: csrc/cuda/embedding_lm_head_sm90.cu [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+                "token_ids and weight must be CUDA tensors");
+    TORCH_CHECK(token_ids.device() == weight.device(),
+                "token_ids and weight must be on the same CUDA device");
+    TORCH_CHECK(weight.dim() == 2, "embedding weight must be [vocab, hidden]");
+    TORCH_CHECK(weight.is_contiguous(), "embedding weight must be contiguous");
+    TORCH_CHECK(is_supported_float_dtype(weight.scalar_type()),
+                "embedding_sm90 supports fp32, fp16, and bf16 weights");
+
+    c10::cuda::CUDAGuard device_guard(weight.device());
+    check_sm90_device();
```
Comment (KJLdefeated) [inline csrc/cuda/embedding_lm_head_sm90.cu:188]: `check_sm90_device()` calls `cudaGetDeviceProperties` on every forward. The Python `_can_use_sm90` already gates on _is_hopper, then the C++ re-checks prop.major == 9 and hard-TORCH_CHECKs. Maybe this is not necessary?

File: csrc/cuda/embedding_lm_head_sm90.cu [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/cuda/CUDAException.h>
+#include <cuda_runtime.h>
+#include <limits>
+#include <torch/extension.h>
+#include <vector>
+
+namespace {
+
+constexpr int kThreads = 256;
```
Comment (KJLdefeated) [inline csrc/cuda/embedding_lm_head_sm90.cu:23]: Non-blocking. One CTA per logit and `kThreads = 256` would be slow for Qwen3 prefill (say 4096 tokens × 151936 vocab) that's ~622M blocks, each reducing 4096 elements with 256 threads (16 elements/thread). But this is fine for reference kernel.

**Review by KJLdefeated — COMMENTED** [review]
Overall look good and clean. Left some comment. After addressed, I am happy to approve.

**Review by maxiaosong1124 — APPROVED** [review]
LGTM now! Thanks!

**Comment by inaniloquentee** [comment]
> Overall look good and clean. Left some comment. After addressed, I am happy to approve.

Thanks! Addressed the review, cached the SM90 device capability check, reduced embedding backward launches with a fixed-order segmented fold, and fixed the LM-head backward test expectation. CI is green now, including H100 SM90.

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thank you for work.
